crates/wit-parser/src/sizealign.rs

```diff
@@ -1,26 +1,210 @@
+use std::{
+    num::NonZeroUsize,
+    ops::{Add, AddAssign},
+};
+
 use crate::{FlagsRepr, Int, Resolve, Type, TypeDef, TypeDefKind};
 
-#[derive(Default)]
-pub enum AddressSize {
-    #[default]
-    Wasm32,
-    Wasm64,
+/// Architecture specific alignment
+#[derive(Eq, PartialEq, PartialOrd, Clone, Copy, Debug)]
+pub enum Alignment {
+    Pointer,
+    Bytes(NonZeroUsize),
 }
 
-#[derive(Default)]
-pub struct SizeAlign {
-    map: Vec<(usize, usize)>,
-    wasm_type: AddressSize,
+impl Default for Alignment {
+    fn default() -> Self {
+        Alignment::Bytes(NonZeroUsize::new(1).unwrap())
+    }
+}
+
+impl std::fmt::Display for Alignment {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Alignment::Pointer => f.write_str("ptr"),
+            Alignment::Bytes(b) => f.write_fmt(format_args!("{}", b.get())),
+        }
+    }
+}
+
+impl Ord for Alignment {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match (self, other) {
+            (Alignment::Pointer, Alignment::Pointer) => std::cmp::Ordering::Equal,
+            (Alignment::Pointer, Alignment::Bytes(b)) => {
+                if b.get() > 4 {
+                    std::cmp::Ordering::Less
+                } else {
+                    std::cmp::Ordering::Greater
+                }
+            }
+            (Alignment::Bytes(b), Alignment::Pointer) => {
+                if b.get() > 4 {
+                    std::cmp::Ordering::Greater
+                } else {
+                    std::cmp::Ordering::Less
+                }
+            }
+            (Alignment::Bytes(a), Alignment::Bytes(b)) => a.cmp(b),
+        }
+    }
+}
+
+impl Alignment {
+    /// for easy migration this gives you the value for wasm32
+    pub fn align_wasm32(&self) -> usize {
+        match self {
+            Alignment::Pointer => 4,
+            Alignment::Bytes(bytes) => bytes.get(),
+        }
+    }
+
+    pub fn format(&self, ptrsize_expr: &str) -> String {
+        match self {
+            Alignment::Pointer => ptrsize_expr.into(),
+            Alignment::Bytes(bytes) => format!("{}", bytes.get()),
+        }
+    }
+}
+
+/// Architecture specific measurement of position,
+/// the combined amount in bytes is
+/// `bytes + if 4 < core::mem::size_of::<*const u8>() { add_for_64bit } else { 0 }`
+#[derive(Default, Clone, Copy, Eq, PartialEq, Debug)]
+pub struct ArchitectureSize {
+    /// exact value for 32-bit pointers
+    pub bytes: usize,
+    /// amount of bytes to add for 64-bit architecture
+    pub add_for_64bit: usize,
+}
```

Comment:
Would it be possible to represent this as something like `bytes` and `pointers`? That way the size could be `bytes + pointers * sizeof(pointer)`. That I think may also make methods like `pointers_to_add` a bit more clear how to be implemented and/or unnecessary since it's just a field then.

Reply:
I tried this once, encountered some difficulties and changed it to the current form. I will try again … I guess the current representation avoids multiplications and divisions and makes alignment more straightforward.

Reply:
In theory it should be possible to skip mul/div by only having shifts left-and-right by storing the log2 of the pointer size (or something like that), but I'm mostly curious about this from an understandability point of view.

Reply:
Oh, while implementing a more obscure test case I realized that I made a conceptual oversight:

I you take the max of `10 bytes` and `2 pointers` you get `4 bytes + 1.5 pointers` - and I didn't take fractional pointer amounts into account, yet. It still can work if I multiply the number of pointers by 4 and handle fractional amounts properly. This wasn't visible in the fuzzing because of the additional alignment applied to the result before outputting the size.

Reply:
Previously this example resulted in `10 bytes + 6 bytes_on_64_bit` - if you start with 9 bytes you even get a quarter pointer amount (9 + 7).

Reply:
Probably you don't want to express/represent partial pointer sizes for objects containing pointers, so you would want to align in the max function and return (8+1).

Reply:
I was able to make it work, but align is a bit more complex than I would like it to be and max now uses the easy way out (calculate 32bit and 64bit then max and then separate again). Max can probably be made more efficient by expanding the 32bit/64bit calculation and then simplifying terms, but as it is mostly used for variants it should be less performance sensitive.

Reply:
Thanks for pushing on this! I think at least from a code generator perspective it'll be easier to reason about size + pointers so I'm glad this was able to work out.